### ISSUE TITLE
Only the part after the last advertisement is played.

### ISSUE THREAD

**Issue by michaelarnauts** [michaelarnauts/plugin.video.vtm.go#8]
Some programs like "Wat Als?" season 3, don't start in the beginning, but it seems like they start from the point after the first commercial break.

**Comment by mediaminister** [comment]
Look at the Period duration, inputstream.adaptive always plays the last period. For Married with children (S09E22) the last Period is 22 minutes, the duration of a full episode. First and second period is 10 seconds and 3 seconds long. This is a Coca-Cola commercial and a VTM GO banner.

**Comment by michaelarnauts** [comment]
Okay. Makes sense. I'm also wondering how this should be handled. Should all periods be merged into one stream? I assume the vtt subtitle position is depending on the episode, and the duration of the ads aren't taken into account in the subtitle timings.

**Comment by mediaminister** [comment]
For the end user it should look like one continuous stream, so all periods should be played in succession. You can find more info about Periods is in the DASH-IF guidelines: https://dashif.org/guidelines/

Currently, I have no clue how to implement this in inputstream.adaptive.
The combination of advertisements and subtitles does not seem easy to implement because this is not part of the MPEG-DASH manifest. VTM GO (Anvato streaming platform) uses a separate json for add cues.

**Comment by michaelarnauts** [comment]
You are right. The VTM GO app probably knows what periods contain the real episode, and adapts the timing of the subtitle to this. We should probably rewrite the vtt ourselves based on the cues.

**Comment by Forceflow** [comment]
Trying to understand this blocking issue:

- Are all periods listed in the MPD file? (as in: are they all known before streaming commences)
- Is there some logic to which periods are Widevine-protected and which aren't? I'm guessing program content has protection, and ads don't?

**Comment by mediaminister** [comment]
> Are all periods listed in the MPD file?

Yes. [vtmvod.txt](https://github.com/michaelarnauts/plugin.video.vtm.go/files/3572591/vtmvod.txt)

> Is there some logic to which periods are Widevine-protected and which aren't?

Yes, each Representation has a ContentProtection tag

> Would a very ugly fix be to parse the MPD file, and list the larger periods (as in: the ones which most likely aren't ads due to their length) as individual kodi media items? So you would end up with FooProgramName (part 1), FooProgramName (part2), etc ... ?
This can only be done in InputStream Adaptive add-on because all video and audio streams are segmented and each individual segment is encrypted: https://github.com/peak3d/inputstream.adaptive/issues/287

**Comment by dagwieers** [comment]
With the latest master branch of **inputstream.adaptive** everything in VTM GO plays fine, including the commercials.

The known issues at this point are:
- subtitles continue through commercials and so the offset is wrong because of commercials
- during a transition from one period to the next (usually when switching from and to commercials) the video stalls briefly, and when video continues, the audio is gone briefly
- proper chapter support for streams is part of Kodi 19 (Matrix) and hopefully we will also get a python interface so we can add chapter support for TV shows in VRT NU as well (so you can jump from topic to topic in a Het Journaal, or De 7de Dag)

So very good news. I think we are very close to releasing VTM GO when the new **inputstream.adaptive** hits all Kodi 18 platforms, even with the above shortcomings.

**Comment by michaelarnauts** [comment]
I agree. Subtitles is not critical to block the first release.

**Comment by dagwieers** [comment]
Only blocker is Linux/arm support at the moment. (LibreELEC and others)

**Comment by peno64** [comment]
I tried this on windows x64 with the latest kodi 19 and inputstream.adaptive 2.5.0 and I still doesn't start from the beginning...

**Comment by mediaminister** [comment]
@peno64 The fix is not yet released and is only available from the master branch: https://jenkins.kodi.tv/blue/organizations/jenkins/peak3d%2Finputstream.adaptive/detail/master/110/artifacts

**Comment by peno64** [comment]
@mediaminister hm, that is a 2.4.0 version and I already have a 2.5.0 version installed. I will wait until it is officially released...

**Comment by dagwieers** [comment]
@peno64 I agree it is pretty confusing that the automated builds have a lower version than what is available on your system. On LibreELEC we still have **inputstream.adaptive 2.3.22**. We are still waiting for VRT NU on a release for something that was fixed 4 months ago...

**Comment by peno64** [comment]
Totday an update of inputstream.adpative was released on the raspberry pi. It is now version 2.4.2.1. I also installed the latest vtm go plugin from release (plugin.video.vtm.go-0.9.0-master-05ebfc8.zip). But still it does not start from the beginning. The subtitles start from the beginning, but the movie not...
And to be sure, I did a reboot of the pi. So the problem still seems to be there. I tried with the movie I, Frankenstein

**Comment by peno64** [comment]
Or is the 2.5 version of inputstream adaptive needed for this to work?

**Comment by mediaminister** [comment]
The fix is not yet released for RPi, you have to build the inputstream adaptive master branch yourself to make it work!

If you can't build this yourself, there are compiled versions available for a limited set of platforms at https://jenkins.kodi.tv/blue/organizations/jenkins/peak3d%2Finputstream.adaptive/detail/master/112/artifacts

**Comment by peno64** [comment]
Unfortunately the jenkins site does not have a rpi version. 
Do you know where to find the instructions to build it myself for the rpi?

**Comment by mediaminister** [comment]
Okay, it seems rather easy to cross compile this from a Linux pc. You need Raspberry Pi tools and firmware: https://github.com/xbmc/xbmc/blob/master/docs/README.RaspberryPi.md

And this are standard instructions to compile Inputstream Adaptive for Kodi 18 Leia:
```
git clone -b Leia https://github.com/xbmc/xbmc.git
git clone https://github.com/peak3d/inputstream.adaptive
cd ~/inputstream.adaptive && mkdir build && cd build
cmake -DADDONS_TO_BUILD=inputstream.adaptive -DADDON_SRC_PREFIX=../.. -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=../../xbmc/addons -DPACKAGE_ZIP=1 ../../xbmc/cmake/addons
make
cd ~/xbmc/addons/
zip -r inputstream.adaptive.zip inputstream.adaptive
```
I will try this myself and come up with detailed instructions and a test zip for RPi if this works out.

**Comment by peno64** [comment]
@mediaminister 
I followed your instructions on a raspberian system on my pi4 and succeeded in generating the inputstream adaptive addon without any errors.
Then I could successfully the generated addon on that pi4 booted with libreelec.
Then when I start a vtm go program which has commercials it now starts with a commercial and then it wants to start the beginning of the program so that is good, but then it either stops working or it starts again with the commercial getting into a loop...
I have to try if something else like vrt.nu or retrospect is still working but can't do this at this moment. Will let know if that still works

**Comment by peno64** [comment]
So I tried this again. When I just play something, for example I Frankenstein, then it starts with a commercial and then it shows a VTM Go advertise and then it wants to play the movie but after a couple of seconds it restarts and plays the commercial again and so on and on.
However if I press pause just at the moment that the movie starts and I wait a couple of seconds and then continue then the movie plays...
Note that scrolling foreward now also works which proofs I do use the latest inputstream adaptive.
Playing a kids program without commercial from vtm go (kabouter plop) or something from the addons retrospect or vrt.nu (het journaal) plays without a problem with this latest inputstream adaptive.
I should now try to run that movie from vtm go with the pause so that it continues and see what happens when it reaches the next commercial break...

**Comment by peno64** [comment]
Ok, so I tried this with the series "locked up". Same scenario. It starts with a commercial and then when the movie starts it restarts. But if i pause for a couple of seconds and then continue it starts to play and then I let it play until the commercial and after the commercial it just continues to play. 
So the problem is only at the start of the movie. But the pause work-around is a small trick that solves this problem. Nobody else having this problem?

**Comment by michaelarnauts** [comment]
@peno64 on Windows and Linux x64, it seems to work fine. I haven't tested yet on arm.

**Comment by peno64** [comment]
​I compiled the latest inputstream.adaptive for libreelec on raspberry pi myself.
You can find it here: http://users.telenet.be/peno/kodi/addons/inputstream.adaptive/

Note that this version is 2.4.0

And believe me, that is the latest version. So make sure that your addons aren't updated automatically or it will be overwritten by an auto update of the kodi addons​

**Comment by michaelarnauts** [comment]
I've bundled all inputstream.adaptive issues that are already fixed upstream, but not yet released in issue #101.
